### ISSUE TITLE
r10k postrun command can keep puppet agent run from completing

### ISSUE THREAD

**Issue by esalberg** [acidprime/r10k#152]
(I opened this ticket with Puppet Support as well, and they directed me here to open an issue in parallel.)

My environment: PE 3.7.2 split install with two additional compile master / one additional mco hub, all on RHEL6.5 with r10k module 2.6.4.  The master-of-masters is puppetca-test.  Deploy keys are configured to our local Gitlab server, and we also use a number of modules from Github.

I noticed that my Puppet masters in my test environment had become listed as Unresponsive in the Puppet Console. Upon attempting to run "puppet agent -t", Puppet reported:

Notice: Run of Puppet configuration client already in progress; skipping (/var/opt/lib/pe-puppet/state/agent_catalog_run.lock exists)

Looking at the process list (ps -ef | grep puppet) on the puppetca-test server, the Agent run had been attempting to complete since April 08 (it is currently April 14):

root 6498 19687 0 Apr08 ? 00:00:21 puppet agent: applying configuration

Based on other hung processes, I'm assuming that the hang occurred while running r10k, as the master is set to do as a postrun command: 
root 7723 6498 0 Apr08 ? 00:00:02 /opt/puppet/bin/ruby /usr/bin/r10k deploy environment -p 
root 10359 7723 0 Apr08 ? 00:00:00 git --git-dir /var/cache/r10k/https---github.com-stahnma-puppet-module-epel.git fetch --prune origin 
root 10362 10359 0 Apr08 ? 00:00:00 git-remote-https origin https://github.com/stahnma/puppet-module-epel.git 
root 10369 10362 0 Apr08 ? 00:00:00 git fetch-pack --stateless-rpc --lock-pack --include-tag --thin --no-progress https://github.com/stahnma/puppet-module-epel.git efs/pull/35/head efs/pull/35/merge

On a second server there is a similar situation, just hanging since April 13 (compile master): 
root 21929 11903 0 Apr13 ? 00:00:17 puppet agent: applying configuration 
root 22964 21929 0 Apr13 ? 00:00:00 /opt/puppet/bin/ruby /usr/bin/r10k deploy environment -p 
root 23007 22964 0 Apr13 ? 00:00:00 git --git-dir /var/cache/r10k/https---github.com-puppetlabs-puppetlabs-concat.git fetch --prune origin 
root 23010 23007 0 Apr13 ? 00:00:00 git-remote-https origin https://github.com/puppetlabs/puppetlabs-concat.git 
root 23017 23010 0 Apr13 ? 00:00:00 git fetch-pack --stateless-rpc --lock-pack --include-tag --thin --no-progress https://github.com/puppetlabs/puppetlabs-concat.git efs/pull/292/head efs/pull/292/merge

I had an issue open with r10k in Puppet 3.3 related to git processes hanging forever if they couldn't complete / somehow got stuck (to the point that I wrote a script to run out of cron to kill r10k git processes if they had been running more than 15 minutes). These hangs occur much less frequently with Puppet 3.7.2 / r10k module 2.6.4 / deploy keys to connect to our local Gitlab over ssh, but apparently they can still happen.

The hangs may be related either to a new feature branch being created or to a slow connection to get a git module (in this case, it could be related to  Github connectivity, e.g. due to a netowrk blip or Github being slow to respond).

(I have also been having some issue with mcollective / the r10k webhook not being acknowledged consistently, but due to the agent running here, I think this is separate from mco / the webhook.)

I can fix the issue by killing the hanging processes and running "r10k deploy environment -pv" by hand (that rarely hangs - even if there's a delay, it almost always carries on and finishes). However, I can leave the test environment in this state if it's helpful for someone to see, especially since I cannot reproduce this condition at will.

My basic thought is that even if there is some issue with the r10k deploy / git fetches, it should not cause the agent run to hang indefinitely, and it ideally also should not cause the git calls / r10k deploy to hang indefinitely.

I also noticed that have been two r10k module releases since I upgraded / one r10k gem update - I will go take a look at the release notes and can update to the latest version (but I won't do so until I know if it's useful for anyone to see the current hanging state).

**Comment by acidprime** [comment]
@esalberg Hmm honestly I am not sure if I know how I can help, fix this in the module though. That postrun_command is mostly a trick to tie the two runs together so you can see failures in r10k in your reporting engine ( though I think that worked better with `prerun_command` it actually causes puppet runs to be gated by r10k so I deprecated its use ).

I would switch to a cron job in the interim and see if you still have the issues, if you do then its an r10k issue and you need to file it [here](https://github.com/puppetlabs/r10k/issues). If it goes away then its either a r10k bug being run via that shell environment or a puppet bug running prerun_commands.

**Comment by acidprime** [comment]
@esalberg RE: mco please file a separate bug on that if you have reproducible behavior

**Comment by acidprime** [comment]
@esalberg I have a little more detail on the intended usage for the postrun covered in this ticket #151 if it helps.

**Comment by adrienthebo** [comment]
@esalberg in addition, try upgrading to r10k 1.5.1 - in 1.4.x and earlier there was a small chance that processes executed by r10k could [deadlock](https://tickets.puppetlabs.com/browse/RK-11) that has been fixed in 1.5.0 and greater. The fact that in [this issue](https://github.com/puppetlabs/r10k/issues/224) reported that the issue didn't manifest itself when you ran in the foreground indicates that this might not be the issue, but it can't hurt.

**Comment by esalberg** [comment]
Thanks - I will go ahead and upgrade and monitor the situation further.

In case it's interesting (mostly for @adrienthebo - I was cleaning up a compile master and killed the git processes that were hanging without killing the r10k deploy process.  It looks like once those git processes were gone, the r10k deploy continued to completion.  Would it be worthwhile for me to open a ticket on r10k itself asking if there's a way r10k could handle a situation where a git process hangs more gracefully, or is that somethig that has been addressed as part of 1.5.x?  I might be able to improve my local Gitlab server, but Github can go through periods of spotty connectivity.
